### ISSUE TITLE
Suggestion: nl_func_type_name_class=ignore should not fallback to nl_func_type_name

### ISSUE THREAD

**Issue by arneforlie** [uncrustify/uncrustify#1700]
In my opinion it is unfortunate that there is no way to explicitly state that formatting according to nl_func_type_name_class shall be ignored when 'nl_func_type_name = add' is used.
I.e. there is no way to avoid that 'nl_func_type_name_class=ignore' value doesn't fallback to value of nl_func_type_name.
The current behaviour is as documented in the configuration file, but I would propose that you either change it, or add a new value to indicate "I really want to ignore this formatting".

As we see from the test case, not only is the return type in declaration on a separate line, but also attributes such as explicit and virtual are affected.

I have tried setting nl_func_type_name_class to 'remove' and 'force' but they don't give me the wanted result of "I really want ignore this". 

Test case
```
#include <string>

class Foo
{
   public:
      explicit Foo(int);
      virtual ~Foo() = default;
      
      virtual void x();
      
      void
      aFunctionWhereIExplicitlyWantNewlineAfterReturnTypeInDeclarationAndDefinition();
   
   private:
      const int i_;
};

Foo::Foo(int i)
   : i_(i)
{
}

void Foo::x()
{
}

void aFunctionWhereIExplicitlyWantNewlineAfterReturnTypeInDeclarationAndDefinition()
{
}

```
Generated result
```
#include <string>

class Foo
{
public:
explicit
Foo(int);
virtual
~Foo() = default;

virtual void x();

void
aFunctionWhereIExplicitlyWantNewlineAfterReturnTypeInDeclarationAndDefinition();

private:
const int i_;
};

Foo::Foo(int i)
	: i_(i)
{
}

void
Foo::x()
{
}

void
aFunctionWhereIExplicitlyWantNewlineAfterReturnTypeInDeclarationAndDefinition()
{
}
```
Wanted result
```
#include <string>

class Foo
{
public:
explicit Foo(int);
virtual ~Foo() = default;

virtual void x();

void
aFunctionWhereIExplicitlyWantNewlineAfterReturnTypeInDeclarationAndDefinition();

private:
const int i_;
};

Foo::Foo(int i)
	: i_(i)
{
}

void
Foo::x()
{
}

void
aFunctionWhereIExplicitlyWantNewlineAfterReturnTypeInDeclarationAndDefinition()
{
}
```
[debug.txt]https://pastebin.com/ExmwKBWG

I have currently hacked my version of Uncrustify to not fallback. It solves my problem, but, of course, is no longer is in accordance with documentation.
```
--- a/src/newlines.cpp  Sat Mar 24 21:53:25 2018 +0100
+++ b/src/newlines.cpp  Tue Mar 27 20:08:39 2018 +0200
@@ -2307,8 +2307,7 @@
             argval_t a = (tmp->parent_type == CT_FUNC_PROTO) ?
                          cpd.settings[UO_nl_func_proto_type_name].a :
                          cpd.settings[UO_nl_func_type_name].a;
-            if (  (tmp->flags & PCF_IN_CLASS)
-               && (cpd.settings[UO_nl_func_type_name_class].a != AV_IGNORE))
+            if ((tmp->flags & PCF_IN_CLASS))
             {
                a = cpd.settings[UO_nl_func_type_name_class].a;
             }
```

**Comment by mwoehlke-kitware** [comment]
> As we see from the test case, not only is the return type in declaration on a separate line, but also attributes such as explicit and virtual are affected.

I think this is fixed by #2144? (Eh, maybe not, but it sounds similar to a bug that #2144 addresses.)

In general, though, I agree. In fact, I was just about to open an issue on the exact same subject, when github helpfully informed me that there was an existing, similar issue :slightly_smiling_face:.

I can think of a couple options (it would be nice if other maintainer(s) could chime in with feedback!):
1. Just make `nl_func_type_name_class` never fall back. But this could surprise folks that expect the current behavior, and doesn't provide any way to set different behavior for declarations vs. definitions. (Okay, but we don't have that today, either.)
2. Deprecate `nl_func_type_name_class` and add `nl_func_def_type_name_class` `nl_func_proto_type_name_class` which never fall back. This also has "surprise" potential, but at least users can get the same behavior as today, as well as the behavior we want.
3. Add an option (`use_nl_func_type_name_class_fallback`?) to control whether `nl_func_type_name_class` falls back. This can maintain compatibility (i.e. default to on, users that don't turn it off will see no change) but doesn't solve the problem of different behavior for definitions vs. declarations.

Personally, I'd prefer (2), especially as I think #1815 gives us a plausible way to maintain compatibility (in which case "deprecate" would definitely become "remove").

**Comment by arneforlie** [comment]
As far as I can see, my original problem has been fixed by #2144 so I will close this issue.
